src/vm/jitinterface.cpp

```diff
@@ -7993,57 +7994,91 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
     }
 
 #ifdef PROFILING_SUPPORTED
-    if (CORProfilerPresent())
     {
-        // #rejit
-        // 
-        // Currently the rejit path is the only path which sets this.
-        // If we get more reasons to set this then we may need to change
-        // the failure reason message or disambiguate them.
-        if (!m_allowInlining)
-        {
-            result = INLINE_FAIL;
-            szFailReason = "ReJIT request disabled inlining from caller";
-            goto exit;
-        }
-
-        // If the profiler has set a mask preventing inlining, always return
-        // false to the jit.
-        if (CORProfilerDisableInlining())
-        {
-            result = INLINE_FAIL;
-            szFailReason = "Profiler disabled inlining globally";
-            goto exit;
-        }
+#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+        // Need to hold the ReJIT lock between when we check if a method should be inlined and
+        // when we report the inlining or there could be a race condition where a method is ReJITted after
+        // checking if it has default IL and before we report the inlining.
+        CrstHolder ch(&(ReJitManager::s_csGlobalRequest));
```

Comment:
Taking a global lock here does not look good. We try to make the whole type loader / JIT process lock free.

Reply:
This caught my eye too in an earlier review, but I wasn't aware there was any official stance on it as long as overall performance testing had satisfactory results.

If I recall correctly I am probably guilty of taking the CodeVersionManager lock in Prestub path in order to determine which version of code should be invoked/jitted. The tiered compilation call-counter also has a lock around it. Not to say that doing a bad thing elsewhere should justify repeating it, but I at least had been undermining it and others perhaps are as well?

This particular case probably isn't that hard to eliminate the lock if we need to.

Reply:
>  it as long as overall performance testing had satisfactory results.

It is not just about performance. It is also about correctness. For example, the correctness of this lock depends on a fragile assumption that the JIT asks for IL before calling `canInline`. It happens to be the case today, but there is nothing that guarantees it. The JIT/EE interface is meant to be order independent. The JIT is free to ask questions in any order it wants and still get correct answers.